### PR TITLE
Add Debuggability to Check Model Weights

### PR DESCRIPTION
**Pull request by rayankrish**
- Improves test code flow by adding a function that handles model preparation.
- Creates the debug.py file which currently contains a weight comparison function and will be populated with more methods in the future.
- Writes a test of the weight comparison and deterministic run.

### CODE REVIEW

File: orttraining/orttraining/python/training/debug.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import os
+import sys
+import torch
+
+from numpy.testing import assert_allclose
+
+# Inputs are the state_dicts for each of the model (dictionary for each layer and weights)
+def compare_onnx_weights(model_a, model_b, verbose=False, rtol=1e-4):
+    r"""Compare onnx weights method
+    Compares the weights of two different ONNX models and throws an error when they diverge
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/debug.py:12]: empty line between the first and second line

File: orttraining/orttraining/python/training/debug.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+
+import numpy as np
+import os
+import sys
+import torch
+
+from numpy.testing import assert_allclose
+
+# Inputs are the state_dicts for each of the model (dictionary for each layer and weights)
+def compare_onnx_weights(model_a, model_b, verbose=False, rtol=1e-4):
```
Comment (rayankrish) [inline orttraining/orttraining/python/training/debug.py:10]: Write this method so it takes two ORTTrainer instances directly rather than expecting the user to pass trainer._training_session.get_state()

File: orttraining/orttraining/python/training/debug.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    r"""Compare onnx weights method
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/debug.py:11]:
```suggestion
    r"""Check whether weights between 'model_a' and 'model_b' ONNX models are up to a certain tolerance 'rtol'
```

File: orttraining/orttraining/python/training/debug.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    Compares the weights of two different ONNX models and throws an error when they diverge
+    Args:
+        model_a, model_b: The state_dicts of two ONNX models containing a dictionary 
+    of the weights for each layer of the respective model.
+        verbose: Indicates if the max absolute difference for each layer should be
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/debug.py:16]:
```suggestion
        verbose (bool, default is False): Indicates if the max absolute difference for each layer should be
```

File: orttraining/orttraining/python/training/debug.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    calculated and printed for debug information. Default: False.
+        rtol: Tolerance for divergence. Default: 1e-4.
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/debug.py:18]:
```suggestion
        rtol (float, default is 1e-4): Tolerance for divergence.
```

File: orttraining/orttraining/python/training/debug.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import torch
+
+from numpy.testing import assert_allclose
+
+# Inputs are the state_dicts for each of the model (dictionary for each layer and weights)
+def compare_onnx_weights(model_a, model_b, verbose=False, rtol=1e-4):
+    r"""Compare onnx weights method
+    Compares the weights of two different ONNX models and throws an error when they diverge
+    Args:
+        model_a, model_b: The state_dicts of two ONNX models containing a dictionary
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/debug.py:14]:
```suggestion
        model_a, model_b (dict): The state_dicts of two ONNX models containing a dictionary 
```

File: orttraining/orttraining/python/training/debug.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    """
+    assert isinstance(model_a, dict) and isinstance(model_b, dict)
+
+    for (a_name, a_val), (b_name, b_val) in zip(model_a.items(), model_b.items()):
+        np_a_vals = np.array(a_val).flatten()
+        np_b_vals = np.array(b_val).flatten()
+        assert np_a_vals.shape == np_b_vals.shape
+        if verbose:
+            print(a_name, np.abs(np_a_vals-np_b_vals).max())
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/debug.py:28]:
```suggestion
            print(f'Weight name: {a_name}: absolute difference: {np.abs(np_a_vals-np_b_vals).max()}')
```

File: orttraining/orttraining/python/training/debug.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+        assert_allclose(a_val, b_val, rtol=rtol, err_msg="weight mismatch")
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/debug.py:29]:
```suggestion
        assert_allclose(a_val, b_val, rtol=rtol, err_msg=f"Weight mismatch for {a_name}")
```

File: orttraining/orttraining/python/training/debug.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    Args:
+        model_a, model_b: The state_dicts of two ONNX models containing a dictionary 
+    of the weights for each layer of the respective model.
+        verbose: Indicates if the max absolute difference for each layer should be
+    calculated and printed for debug information. Default: False.
+        rtol: Tolerance for divergence. Default: 1e-4.
+
+    """
+    assert isinstance(model_a, dict) and isinstance(model_b, dict)
+
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/debug.py:22]: Should the dict have the same size? Any other property that we should *assert* before testing?

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 10]
```diff
@@ -456,20 +457,7 @@ def testLRSchedulerUpdateImpl(lr_scheduler, expected_values):
...
-                                                    0.023843, 0.023843, 0.023843, 0.023843, 0.023843]),
-    ('train_step', optim.lr_scheduler.CosineWarmupLRScheduler, [0.181818, 0.066116, 0.036063, 0.026228, 0.023843,
-                                                  0.010225, 0.002989, 0.0005158, 0.000040937, 0.0000008291]),
-    ('train_step', optim.lr_scheduler.LinearWarmupLRScheduler, [0.181818, 0.066116, 0.036063, 0.026228, 0.023843,
-                                                  0.021675, 0.0157636, 0.0085983, 0.0031266, 0.00056847]),
-    ('train_step', optim.lr_scheduler.PolyWarmupLRScheduler, [0.181818, 0.066116, 0.036063, 0.026228, 0.023843,
-                                                0.0160749, 0.0096935, 0.0050622, 0.0021585, 0.000650833])
-])
-def testInstantiateORTTrainer(step_fn, lr_scheduler, expected_lr_values):
+def generate_pytorch_transformer_model_sample(optim_config, options=None, step_fn='train_step', device='cpu'):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:460]:
```suggestion
def generate_pytorch_transformer_model_sample(optim_config, options={}, step_fn='train_step', device='cpu'):
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 11]
```diff
@@ -502,20 +520,16 @@ def testInstantiateORTTrainer(step_fn, lr_scheduler, expected_lr_values):
         opts.update({'lr_scheduler' : lr_scheduler(max_train_step, warmup)})
    
     opts = orttrainer.ORTTrainerOptions(opts)
-    trainer = orttrainer.ORTTrainer(model, model_desc, optim_config, loss_fn=my_loss, options=opts)
 
-    # Preparing data
-    train_data, val_data, _ = utils.prepare_data('cpu', 20, 20)
+    model, model_desc, trainer, data, targets = generate_pytorch_transformer_model_sample(optim_config, opts, step_fn)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:524]:
```suggestion
        # Using Pytorch Transfor model as example
        model, model_desc, trainer, data, targets = generate_pytorch_transformer_model_sample(optim_config, opts, step_fn)
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 12]
```diff
@@ -573,3 +587,45 @@ def testInstantiateORTTrainer(step_fn, lr_scheduler, expected_lr_values):
...
+@pytest.mark.parametrize("seeds, device_id", [
+    ((0, 0), 'cpu'),
+    ((42, 42), 'cpu'),
+    ((42, 57), 'cpu'),
+    ((0, 0), 'cuda:0'),
+    ((0, 0), 'cuda')
+])
+def testORTDeterministicCompute(seeds, device_id):
+    optim_config = optim.LambConfig()
+    print(torch.device(device_id))
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:600]:
```suggestion
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 13]
```diff
@@ -573,3 +587,45 @@ def testInstantiateORTTrainer(step_fn, lr_scheduler, expected_lr_values):
...
+        'device' : {
+            'id' : device_id,
+            'mem_limit' : 256
+        }
+    })
+    
+    torch.manual_seed(seeds[0])
+    set_seed(seeds[1])
+
+    model, model_desc, trainer, data, targets = generate_pytorch_transformer_model_sample(optim_config, opts,  device=device_id)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:614]:
```suggestion
    # Using Pytorch Transfor model as example
    model, model_desc, trainer, data, targets = generate_pytorch_transformer_model_sample(optim_config, opts,  device=device_id)
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 14]
```diff
@@ -573,3 +587,45 @@ def testInstantiateORTTrainer(step_fn, lr_scheduler, expected_lr_values):
     assert (trainer_from_onnx._onnx_model.graph == trainer._onnx_model.graph)
     assert (onnx.helper.printable_graph(trainer_from_onnx._onnx_model.graph) == onnx.helper.printable_graph(trainer._onnx_model.graph))
 
+
+@pytest.mark.parametrize("seeds, device_id", [
+    ((0, 0), 'cpu'),
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:592]: I would simplify by using just one seed unless there is a scenario we can say this two values can capture a bug

File: orttraining/orttraining/python/training/debug.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+from numpy.testing import assert_allclose
+from onnxruntime.capi.training import orttrainer
+
+def compare_onnx_weights(model_a, model_b, verbose=False, rtol=1e-4):
+    r"""Compare whether weights between 'model_a' and 'model_b' ONNX models are within
+    a certain tolerance 'rtol'
+
+    Compares the weights of two different ONNX models and throws an error when they diverge
+    Args:
+        model_a, model_b (orttrainer): Two instances of ORTTrainer with the same model structure
```
Comment (rayankrish) [inline orttraining/orttraining/python/training/debug.py:16]: Correct case for ORTTrainer

File: orttraining/orttraining/python/training/debug.py [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+        verbose (bool, default is False): Indicates if the max absolute difference for each layer should be
+    calculated and printed for debug information.
+        rtol (float, default is 1e-4): Tolerance for divergence.
+
```
Comment (rayankrish) [inline orttraining/orttraining/python/training/debug.py:20]: Remove line

File: orttraining/orttraining/python/training/debug.py (hunk 15 above)
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/debug.py:16]:
```suggestion
        model_a, model_b (ORTTrainer): Two instances of ORTTrainer with the same model structure
```

File: orttraining/orttraining/python/training/debug.py (hunk 16 above)
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/debug.py:20]:
```suggestion
```

Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/debug.py:20]:
```suggestion
```

File: orttraining/orttraining/python/training/debug.py [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+        np_a_vals = np.array(a_val).flatten()
+        np_b_vals = np.array(b_val).flatten()
+        assert np_a_vals.shape == np_b_vals.shape
+        if verbose:
+            print(f'Weight name: {a_name}: absolute difference: {np.abs(np_a_vals-np_b_vals).max()}')
+        assert_allclose(a_val, b_val, rtol=rtol, err_msg=f"Weight mismatch for {a_name}")
+         
+
+
+
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/debug.py:35]:
```suggestion
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 18]
```diff
@@ -485,10 +473,40 @@ def testInstantiateORTTrainer(step_fn, lr_scheduler, expected_lr_values):
     utils = _utils.import_module_from_file(utils_path, utils_name)
 
     # Modeling
-    model = pt_model.TransformerModel(28785, 200, 2, 200, 2, 0.2)
+    model = pt_model.TransformerModel(28785, 200, 2, 200, 2, 0.2).to(device)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:476]:
```suggestion
    model = pt_model.TransformerModel(28785, 200, 2, 200, 2, 0.2)
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 19]
```diff
@@ -573,3 +588,44 @@ def testInstantiateORTTrainer(step_fn, lr_scheduler, expected_lr_values):
     assert (trainer_from_onnx._onnx_model.graph == trainer._onnx_model.graph)
     assert (onnx.helper.printable_graph(trainer_from_onnx._onnx_model.graph) == onnx.helper.printable_graph(trainer._onnx_model.graph))
 
+
+@pytest.mark.parametrize("seed, device_id", [
+    (0, 'cpu'),
+    (42, 'cpu'),
+    (0, 'cuda:0'),
+    (0, 'cuda')
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:596]:
```suggestion
    (24, 'cuda')
```

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 20]
```diff
@@ -573,3 +588,44 @@ def testInstantiateORTTrainer(step_fn, lr_scheduler, expected_lr_values):
...
+])
+def testORTDeterministicCompute(seed, device_id):
+    optim_config = optim.LambConfig()
+    opts = orttrainer.ORTTrainerOptions({
+        'debug' : {
+            'deterministic_compute': True
+        },
+        'device' : {
+            'id' : device_id,
+            'mem_limit' : 256
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:606]:
```suggestion
            'mem_limit' : 10*1024*1024
```
